alg/teca_temporal_index_select.cxx

```diff
@@ -0,0 +1,292 @@
+#include <cuda_runtime.h>
+#endif
+
+//#define TECA_DEBUG
+
+
+// --------------------------------------------------------------------------
+teca_temporal_index_select::teca_temporal_index_select()
+{
+    this->set_number_of_input_connections(1);
+    this->set_number_of_output_ports(1);
+}
+
+// --------------------------------------------------------------------------
+teca_temporal_index_select::~teca_temporal_index_select()
+{}
+
+#if defined(TECA_HAS_BOOST)
+// --------------------------------------------------------------------------
+void teca_temporal_index_select::get_properties_description(
+    const std::string &prefix, options_description &global_opts)
+{
+    options_description opts("Options for "
+        + (prefix.empty()?"teca_temporal_index_select":prefix));
+
+    opts.add_options()
+        TECA_POPTS_MULTI_GET(std::vector<long long>, prefix, indices,
+            "a list of the time indices to select")
+        ;
+
+    this->teca_algorithm::get_properties_description(prefix, opts);
+
+    global_opts.add(opts);
+}
+
+// --------------------------------------------------------------------------
+void teca_temporal_index_select::set_properties(
+    const std::string &prefix, variables_map &opts)
+{
+    this->teca_algorithm::set_properties(prefix, opts);
+
+    TECA_POPTS_SET(opts, std::vector<long long>, prefix, indices)
+}
+#endif
+
+
+// --------------------------------------------------------------------------
+teca_metadata teca_temporal_index_select::get_output_metadata(unsigned int port,
+    const std::vector<teca_metadata> &md_in)
+{
+#ifdef TECA_DEBUG
+    std::cerr << teca_parallel_id()
+        << "teca_temporal_index_select::get_output_metadata" << std::endl;
+#endif
+    (void) port;
+
+    // copy the incoming metadata
+    teca_metadata md_out = teca_metadata(md_in[0]);
+
+    // skip modifying the coordinates if no indices have been requested
+    if (! (this->indices.size() == 0) )
+    {
+        // revise the time coordinate to be the subset that was requested
+        teca_metadata coords;
+        p_teca_variant_array t_in;
+        if ( md_out.get("coordinates", coords) || !(t_in = coords.get("t")) )
+        {
+            TECA_FATAL_ERROR("failed to locate target mesh coordinates")
+            return md_out;
+        }
+        // get the incoming time values
+        p_teca_variant_array t_out = t_in->new_copy();
+        // create the output times by selecting times at the given indices
+        t_out->resize(this->indices.size());
+        for ( size_t i = 0; i < this->indices.size(); ++i )
+        {
+            NESTED_TEMPLATE_DISPATCH_FP(
+                const teca_variant_array_impl,
+                t_in.get(), 1,
+                auto sp_t_in = dynamic_cast<TT1*>
+                    (t_in.get())->get_cpu_accessible();
+
+                const NT1 *p_t_in = sp_t_in.get();
+                // set this time value
+                t_out->set(i, p_t_in[this->indices[i]]);
```

Comment:
could you please expand on why it would be better not to use the virtual API here and in what cases the virtual API should be used?  I think there's a point about C++ code design that I'm unfamiliar with

Reply:
it's inefficient. the virtual api could be useful when you don't know the type.

The virtual methods on teca_variant_array handles all the casting internally, then routes to the correct device CPU or GPU. so these are the overheads here:  each call to virtual set has the cost of a function call  because virtual calls can't be in lined, then our code has to figure out the actual type, then the allocator, and finally set the value (this may result in another function call eg. cudaMemcpy). you're doing it in a loop so it happens multiple times.

Here you know the type. So instead you can cast to the typed array and use the `data()` method to get the pointer, then each setting each value is a single machine instruction. All of the allocator based routing is skipped. The case can be done outside the loop.

Reply:
as a rule of thumb: use the virtual methods when you can transfer all the data with a single call.

Reply:
one of the sad things about the changes to support GPUs is that calls to set or get a single value that were essentially free before, are now very very costly. Now we need to use bulk data transfers when ever possible.

Reply:
got it, thanks for the thorough explanation!